feat(web-server): Use isbinaryfile for binary file detection

mramato opened the pull request:
Delete `lib/binary-extensions.json` and change the preprocessor to use the `isbinaryfile` module. This is a very simple change but there are some whitespace only changes due to the `isBinaryFile` function being asynchronous and therefore adding an extra level of indentation, append `?w=1` to the PR url to have the diff ignore whitespace.

Closes #1070

GitCop commented:
Thank you for your contribution. We detected an issue with your commit message though and would ask you kindly to fix it.

 * Commit: 12d1085cb45975098e295a0698e6464fdbe6d3fc
  * Your commit message body contains a line that is longer than 80 characters

Guidelines are available at http://karma-runner.github.io/0.13/dev/git-commit-msg.html and please do not hesitate to ask if you have any questions.

---
This message was auto-generated by https://gitcop.com

Dignifiedquire commented on package.json:
```diff
@@ -249,7 +249,8 @@
     "Jurko Gospodnetić <jurko.gospodnetic@pke.hr>",
-    "Karl Lindmark <karl.lindmark@ninetwozero.com>"
+    "Karl Lindmark <karl.lindmark@ninetwozero.com>",
+    "Matthew Amato <matt.amato@gmail.com>"
```
Please revert this, we generate this list automatically on release via the git log

Dignifiedquire commented on lib/preprocessor.js:
```diff
@@ -51,64 +45,69 @@ var createPreprocessor = function (config, basePath, injector) {
-        file.content = content
-        file.sha = sha1(content)
-        return done()
+      var nextPreprocessor = function (error, content) {
```
This nesting is getting pretty deep, can you pull this into a separate function please?

mramato replied:
Just to clarify, you want just `nextPreprocessor` factored into a separate function, or the entire `isBinaryFile` callback?

mramato replied:
I factored out `nextPreprocessor` into it's own function.  I had to change the signature, which I hope is okay.  Otherwise it would have to be a factor function.

Dignifiedquire commented:
Thanks, left some comments inline

mramato commented:
> Thanks, left some comments inline

Thanks, should I squash this PR back into a single commit after making changes?

GitCop commented:
Thank you for your contribution. We detected an issue with your commit message though and would ask you kindly to fix it.

 * Commit: c9b6479fa9e1a8ae69e46cda7478bedb8c40d9a9
  * Commits must be in the following format: **%{type}(?%?{scope})?: %{description}**
  * Your subject line is longer than 70 characters

Guidelines are available at http://karma-runner.github.io/0.13/dev/git-commit-msg.html and please do not hesitate to ask if you have any questions.

---
This message was auto-generated by https://gitcop.com

mramato commented:
All requested changes have been made.  I'm getting one failing local test and I'm not sure why.  Any advice would be appreciated.  Here is the call stack below:

```
  1) preprocessor "before each" hook for "should preprocess matching file":
     TypeError: must start with number, buffer, array or string
    at fromObject (buffer.js:139:11)
    at new Buffer (buffer.js:58:10)
    at Context.<anonymous> (C:\Git\mramato\karma\test\unit\preprocessor.spec.js:24:52)
    at callFn (C:\Git\mramato\karma\node_modules\mocha\lib\runnable.js:286:21)
    at Hook.Runnable.run (C:\Git\mramato\karma\node_modules\mocha\lib\runnable.js:279:7)
    at next (C:\Git\mramato\karma\node_modules\mocha\lib\runner.js:297:10)
    at Immediate._onImmediate (C:\Git\mramato\karma\node_modules\mocha\lib\runner.js:319:5)
```

Dignifiedquire commented on lib/preprocessor.js:
```diff
@@ -1,8 +1,7 @@
 var crypto = require('crypto')
 var mm = require('minimatch')
-var extensions = require('./binary-extensions.json').extensions
+var isBinaryFile = require('isBinaryFile')
```
looks like this needs to be downcased to `require('isbinaryfile')` (see travis failures)

mramato replied:
whoops, thanks.  Stupid Windows case insensitivity.

Dignifiedquire commented:
> Thanks, should I squash this PR back into a single commit after making changes?

Yes for things like this, there should be a single commit when I merge it.

GitCop commented:
Thank you for your contribution. We detected an issue with your commit message though and would ask you kindly to fix it.

 * Commit: c9b6479fa9e1a8ae69e46cda7478bedb8c40d9a9
  * Commits must be in the following format: **%{type}(?%?{scope})?: %{description}**
  * Your subject line is longer than 70 characters

 * Commit: 22699db6b8ff557c8d58de0393afcf019c3504d4
  * Commits must be in the following format: **%{type}(?%?{scope})?: %{description}**

Guidelines are available at http://karma-runner.github.io/0.13/dev/git-commit-msg.html and please do not hesitate to ask if you have any questions.

---
This message was auto-generated by https://gitcop.com

Dignifiedquire commented:
The error comes from you using `new Buffer()` without any content. If you want to create an empty buffer just do sth like `new Buffer(20).fill(0)`

mramato commented:
> The error comes from you using new Buffer() without any content. If you want to create an empty buffer just do sth like new Buffer(20).fill(0)

Thanks, I realized that already now I'm trying to figure out why tests are still failing.  I think there is an issue with the `isbinaryfile` module and the mocking that's going on in the tests.  I'll let you know once I figure it out.

mramato commented:
On an unrelated note, the `validate-commit-msg` step crashes on my system whenever I try to commit.  My only option is to comment out the step in package.json temporarily (which is my the bot yelled at me a few times).  Here's the callstack:

```
C:\Git\mramato\karma\node_modules\validate-commit-msg\index.js:103
      return buffer.toString().split('\n').shift();
                   ^

TypeError: Cannot read property 'toString' of undefined
    at firstLineFromBuffer (C:\Git\mramato\karma\node_modules\validate-commit-msg\index.js:103:20)
    at ReadFileContext.callback (C:\Git\mramato\karma\node_modules\validate-commit-msg\index.js:92:15)
    at FSReqWrap.readFileAfterOpen [as oncomplete] (fs.js:303:13)
```

Should I file a new issue for this?  I'm on Windows 7, Node 4.2.x LTS

Dignifiedquire commented:
please file an issue for this on the validate-commit-msg repo

mramato commented:
All of my changes are in and have been squashed into a single commit.

I just noticed that two tests are failing on Node 0.10 tests (but tests pass on all other versions), but it looks like Node 0.10 fails in karma master too.  I'm not sure why my branch is failing other than possible incompatibility between `isbinaryfile` module and 0.10.  If anyone has any thoughts, I'd appreciate it.  Otherwise, I can try 0.10 under nvm when I'm home on my Linux machine.  I assume 0.10 compatibility is still required?

Dignifiedquire commented:
Yes we need 0.10 compat, but those tests are just time sensitive and flaky on travis for that reason. I've retriggered it and it should be going green soon :)

Dignifiedquire commented:
Hmm looks like this is a real failure on 0.10 now :( if you can take a look later today I'd appreciate it. Otherwise I can check it out tomorrow

mramato commented:
So I ran the tests over 40 times locally under v0.10.41 on Linux Mint 17.3 and everything passes.  I have no idea why they failed on travis.

mramato commented:
Nevermind, I'm a moron and forgot to check out my branch after the initial clone on my linux partition.  The tests fail every time.

mramato commented:
Okay, I've addressed all problems and squashed my latest changes into a single commit.  This is ready for final review (and hopefully merger).  Thanks for the help.

Dignifiedquire commented:
Thanks :octocat: this looks good to me :)
